Update conda channel order to sync with Bioconda

nsoranzo opened the pull request:
Also:
- Remove `_ensure_conda_context()` function and make `conda_context` argument mandatory in functions/methods.
  This should help making sure the correct context (e.g. channels) is used for all Conda operations.
- Remove unused `is_target_available()` function.
-  Override channel list on the command line instead of modifying condarc.
  The way we were modifying condarc in the `ensure_channels_configured()` method was broken because we were appending the missing channels at the end, so the final order may not be the one specified in `conda_ensure_channels` .
  Also not modifying the user condarc is a much better approach and overriding the channel list on the command line will make debugging easier.

nsoranzo committed:
Update conda channel order to sync with Bioconda

bgruening reviewed (approved):
@nsoranzo thanks for doing this! It is somewhere on my todo list. 

I think we also need to update planemo and galaxy-lib.
Do we want to leave this open until we have PRs for above mentioned PRs as well?

nsoranzo commented:
@bgruening Any idea why this integration test fails? https://jenkins.galaxyproject.org/job/docker-integration/1349/testReport/junit/integration.test_resolvers/CondaResolutionIntegrationTestCase/test_legacy_r_mapping/?auto_refresh=false

It seems it's mixing `r-base` from `conda-forge` with the [broken](https://github.com/ContinuumIO/anaconda-recipes/pull/60) `readline-6.2-2` from the `defaults` channel instead of `readline-6.2-0` from `conda-forge`.

nsoranzo commented:
xref. https://github.com/galaxyproject/planemo/pull/715
I think that `galaxy-lib` will be updated when the changes from this `galaxy` repository are merged in it.

bgruening commented:
It looks like most of the packages are coming from default - this is wired. Most of them, if not all, should come from conda-forge.

jmchilton commented:
I'll take care of planemo and galaxy-lib after we merge this - but that test failure is scary. I think I pulled that example from an actual tool that needed the merge Conda installs.

nsoranzo commented:
@jmchilton Should we add a warning at Galaxy startup if the configured channel order is not the one in `conda_ensure_channels` ?

nsoranzo commented:
I think the issue is that the [galaxy.tools.deps.conda_util.CondaContext.ensure_channels_configured()](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/deps/conda_util.py#L147) method appends the configured channels at the end, which by default is the `defaults` channel. So `defaults` remains the highest priority channel instead of being the lowest.

nsoranzo commented:
My plan to fix the problem here is to remove ensure_channels_configured(), load_condarc() and save_condarc() and instead override the channel list on the command line as in https://github.com/galaxyproject/galaxy/blob/98292a0a124a106bde64dd977f0e505e599ff864/lib/galaxy/tools/deps/conda_util.py#L504-L506

Please let me know if you have reservations.

mvdbeek commented:
I think that is the best approach @nsoranzo, no need to manipulate and conflict with users' `.condarc`

nsoranzo committed:
Remove _ensure_conda_context() function

and make conda_context argument mandatory in functions/methods.
This should help making sure the correct context (e.g. channels)
is used for all Conda operations.

Also remove unused is_target_available() function.

nsoranzo committed:
Override channel list on the command line instead of modifying condarc

The way we were modifying condarc in the ensure_channels_configured()
method was broken because we were appending the missing channels at the
end, so the final order may not be the one specified in
conda_ensure_channels .

Also:
- not modifying the user condarc is a much better approach
- overriding the channel list on the command line will make debugging
  easier
